### PR TITLE
Custom gradient for ``keras`` interface

### PR DESCRIPTION
**Pull request by BrunoLiegiBastonLiegi**
This implements custom gradient calculation for the ``keras`` interface. In particular making it possible to train a model using the ``PSR`` or ``Jax`` differentiation.

### CODE REVIEW

**Review by MatteoRobbiati — COMMENTED** [review]
Thanks for this! I leave some comments below :)

File: src/qiboml/interfaces/keras.py [hunk 1]
```diff
@@ -15,12 +19,19 @@
...
     "numpy": Jax,
 }
 
+QIBO_2_KERAS_BACKEND = {
+    "qiboml (pytorch)": "torch",
+    "qiboml (tensorflow)": "tensorflow",
+    "qiboml (jax)": "jax",
+    "numpy": "numpy",
+}
+
```
[the remark this answers is not in the document]
Comment (BrunoLiegiBastonLiegi) [reply src/qiboml/interfaces/keras.py:34]: Yep indeed, this was here because in principle keras can work directly with the torch and jax backends without going through the custom gradients thing. However, as this requires some reworking of the tests, I preferred to defer this to a future PR, thus to avoid confusion better to remove this indeed.

File: src/qiboml/interfaces/keras.py [hunk 2]
```diff
@@ -53,26 +70,28 @@ def __post_init__(self):
...
-        raise NotImplementedError
-        # return custom_operation(
-        #    self.encoding,
-        #    self.circuit,
-        #    self.decoding,
-        #    self.differentiation,
-        #    self.circuit_parameters,
-        #    x,
-        # )
+        return self.custom_gradient.evaluate(x, 1 * self.circuit_parameters)
```
[the remark this answers is not in the document]
Comment (BrunoLiegiBastonLiegi) [reply src/qiboml/interfaces/keras.py:94]: Because for pytorch the ``forward`` is a static method and all the needed context is passed through the ``ctx`` argument, unfortunately the same is not possible for keras.

File: src/qiboml/interfaces/keras.py [hunk 3]
```diff
@@ -91,3 +110,80 @@ def backend(
...
+
+        def forward(x, params):
+            x = self.backend.cast(
+                keras.ops.convert_to_numpy(x), dtype=self.backend.np.float64
+            )
+            circuit = self.encoding(x) + self.circuit
+            circuit.set_parameters(params)
+            y = self.decoding(circuit)
+            y = self.backend.to_numpy(y)
+            return keras.ops.cast(y, dtype=y.dtype)
```
[the remark this answers is not in the document]
Comment (BrunoLiegiBastonLiegi) [reply src/qiboml/interfaces/keras.py:143]: Because the ``keras.ops.cast`` might still fail in some cases, e.g. if ``y`` is a torch tensor on GPU.

File: tests/test_interfaces.py [hunk 4]
```diff
@@ -227,7 +226,13 @@ def random_parameters(frontend, model):
     elif frontend.__name__ == "qiboml.interfaces.keras":
         new_params = []
         for weight in model.get_weights():
-            new_params += [weight + frontend.tf.random.uniform(weight.shape) / 5]
+            if frontend.keras.backend.backend() == "tensorflow":
+                val = frontend.tf.random.uniform(weight.shape)
+            elif frontend.keras.backend.backend() == "pytorch":
+                val = frontend.torch.randn(weight.shape)
```
[the remark this answers is not in the document]
Comment (BrunoLiegiBastonLiegi) [reply tests/test_interfaces.py:232]: ah no, just forgot what ``randn`` was generating.

File: tests/test_differentiation_rules.py [hunk 5]
```diff
@@ -1,50 +1,74 @@
...
 TARGET_GRAD = np.array([0.130832955241203, 0.0, -1.806316614151001, 0.0])
 TARGET_GRAD = {
     "no_inputs": np.array([0.130832955241203, 0.0, -1.806316614151001, 0.0]),
     "wrt_inputs": np.array([0.740709794, 0.0, -0.730872325, 0.0]),
 }
 
-torch.set_default_dtype(torch.float64)
-torch.set_printoptions(precision=15, sci_mode=False)
+
+def set_seed(frontend, seed):
```
[the remark this answers is not in the document]
Comment (BrunoLiegiBastonLiegi) [reply tests/test_differentiation_rules.py:23]: yep I think it is set separately in the tests directly.
